### PR TITLE
hello-oboe: Allow explicit selection of audio API (OpenSLES or AAudio)

### PR DESCRIPTION
**Pull request by dturner**

### CODE REVIEW

File: oboe/hello-oboe/src/main/cpp/jni_bridge.cpp [hunk 1]
```diff
@@ -53,6 +53,20 @@ Java_com_google_sample_oboe_hellooboe_PlaybackEngine_native_1setToneOn(JNIEnv *e
     engine->setToneOn(isToneOn);
 }
 
+JNIEXPORT void JNICALL
+Java_com_google_sample_oboe_hellooboe_PlaybackEngine_native_1setAudioApi(JNIEnv *env, jclass type,
```
Comment (JoseAlcerreca) [inline oboe/hello-oboe/src/main/cpp/jni_bridge.cpp]: In a proper language you would do
```
Java_com_google_sample_oboe_hellooboe_PlaybackEngine_native_1setAudioApi(
        JNIEnv *env, jclass type,
        jlong engineHandle...) {
```

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java [hunk 2]
```diff
@@ -74,6 +76,26 @@ protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
 
+        mAudioApiSpinner = findViewById(R.id.audioApisSpinner);
+        mAudioApiSpinner.setAdapter(new SimpleAdapter(
+                this,
+                createAudioApisOptionsList(),
+                R.layout.audio_apis_spinner, // the xml layout
+                new String[]{getString(R.string.audio_api_description_key)}, // field to display
```
Comment (JoseAlcerreca) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: nit: `new int[] {5, 6} and new int[] { 5, 6 } are both valid` but space between `[]` and `{` is needed

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java [hunk 3]
```diff
@@ -74,6 +76,26 @@ protected void onCreate(Bundle savedInstanceState) {
...
+                new int[]{R.id.audioApiOption})); // View to show field in
```
Comment (JoseAlcerreca) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: ditto

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java [hunk 4]
```diff
@@ -138,36 +160,37 @@ private int getBufferSizeInBursts(){
 
     private void setupLatencyUpdater() {
 
-        if (PlaybackEngine.isLatencyDetectionSupported()){
+        //Update the latency every 1s
+        TimerTask latencyUpdateTask = new TimerTask() {
+            @Override
+            public void run() {
+
+            final String latencyStr;
```
Comment (JoseAlcerreca) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: indent

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java [hunk 5]
```diff
@@ -138,36 +160,37 @@ private int getBufferSizeInBursts(){
...
 
-            //Update the latency every 1s
-            TimerTask latencyUpdateTask = new TimerTask() {
+            if (PlaybackEngine.isLatencyDetectionSupported()){
```
Comment (JoseAlcerreca) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: space before `{`

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java [hunk 6]
```diff
@@ -138,36 +160,37 @@ private int getBufferSizeInBursts(){
...
+
+                double latency = PlaybackEngine.getCurrentOutputLatencyMillis();
+                if (latency >= 0){
```
Comment (JoseAlcerreca) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: space before {

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java [hunk 7]
```diff
@@ -201,4 +224,17 @@ protected void onDestroy() {
 
         return bufferSizeOptions;
     }
+
+    private List<HashMap<String,String>> createAudioApisOptionsList(){
+
+        ArrayList<HashMap<String,String>> audioApiOptions = new ArrayList<>();
+
+        for (int i = 0; i < AUDIO_API_OPTIONS.length; i++){
```
Comment (JoseAlcerreca) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: space

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/PlaybackEngine.java [hunk 8]
```diff
@@ -43,6 +43,10 @@ static void setToneOn(boolean isToneOn){
         if (mEngineHandle != 0) native_setToneOn(mEngineHandle, isToneOn);
     }
 
+    static void setAudioApi(int audioApi){
```
Comment (JoseAlcerreca) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/PlaybackEngine.java]: check for missing space everywhere

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java [hunk 9]
```diff
@@ -201,4 +224,17 @@ protected void onDestroy() {
...
+            HashMap<String,String> option = new HashMap<>();
```
Comment (JoseAlcerreca) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: space after ,

File: oboe/hello-oboe/src/main/cpp/jni_bridge.cpp (hunk 1 above)
Comment (dturner) [reply oboe/hello-oboe/src/main/cpp/jni_bridge.cpp]: Not sure what you mean. In C++ you can do `void method(param1, ...)` but explicit parameter types are preferable in this case.

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java (hunk 4 above)
Comment (dturner) [reply oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: Done

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java (hunk 6 above)
Comment (dturner) [reply oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: Done

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java (hunk 7 above)
Comment (dturner) [reply oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: Done

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java (hunk 9 above)
Comment (dturner) [reply oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: Done

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/PlaybackEngine.java (hunk 8 above)
Comment (dturner) [reply oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/PlaybackEngine.java]: Done

File: oboe/hello-oboe/src/main/cpp/jni_bridge.cpp (hunk 1 above)
Comment (JoseAlcerreca) [reply oboe/hello-oboe/src/main/cpp/jni_bridge.cpp]: break line after the opening parenthesis and put each parameter in one line.

File: oboe/hello-oboe/src/main/cpp/PlayAudioEngine.cpp [hunk 10]
```diff
@@ -299,3 +300,14 @@ void PlayAudioEngine::setBufferSizeInBursts(int32_t numBursts) {
 bool PlayAudioEngine::isLatencyDetectionSupported() {
     return mIsLatencyDetectionSupported;
 }
+
+void PlayAudioEngine::setAudioApi(oboe::AudioApi audioApi) {
+    if (audioApi != mAudioApi){
```
Comment (philburk) [inline oboe/hello-oboe/src/main/cpp/PlayAudioEngine.cpp]: space before {

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java [hunk 11]
```diff
@@ -74,9 +76,30 @@ protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
 
+        mAudioApiSpinner = findViewById(R.id.audioApisSpinner);
```
Comment (philburk) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: ApisSpinner vs ApiSpinner is confusing

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java [hunk 12]
```diff
@@ -93,11 +116,12 @@ public void onNothingSelected(AdapterView<?> adapterView) {
 
         mBufferSizeSpinner = findViewById(R.id.bufferSizeSpinner);
         mBufferSizeSpinner.setAdapter(new SimpleAdapter(
-                this,
```
Comment (philburk) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: double indent is preferred over single indent for continuing code

File: oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java [hunk 13]
```diff
@@ -138,36 +162,37 @@ private int getBufferSizeInBursts(){
...
+                    latencyStr = getString(R.string.only_supported_on_api_26);
                 }
-            };
 
-            mLatencyUpdater = new Timer();
-            mLatencyUpdater.schedule(latencyUpdateTask, 0, UPDATE_LATENCY_EVERY_MILLIS);
+                runOnUiThread(new Runnable() {
+                    @Override
+                    public void run() {
+                    mLatencyText.setText(getString(R.string.latency, latencyStr));
```
Comment (philburk) [inline oboe/hello-oboe/src/main/java/com/google/sample/oboe/hellooboe/MainActivity.java]: indent

**Comment by dturner** [comment]
Thanks Phil! I'll address these on the master branch since this was merged already.

**Comment by dturner** [comment]
Addressed here: https://github.com/googlesamples/android-audio-high-performance/commit/7f9731f815061b257ee5899baec38f1a85567baa

**Pull request merged**